### ISSUE TITLE
Localisation: Turkish

### ISSUE THREAD

**Issue by JuliusSweetland** [OptiKey/OptiKey#206]
**1.Transifex translation of UI strings:** Complete

**2.Keyboard layout:**
![image](https://cloud.githubusercontent.com/assets/4722561/13409180/a900814e-df29-11e5-979f-0ae9accb4272.png)
http://ascii-table.com/keyboard.php/179

**3.Alphabet:**
a)Letters: 29 individual letters - https://en.wikipedia.org/wiki/Turkish_alphabet#Letters
Capital letters
A	B	C	Ç	D	E	F	G	Ğ	H	İ	I	J	K	L	M	N	O	Ö	P	R	S	Ş	T	U	Ü	V	Y	Z
Lower case letters
a	b	c	ç	d	e	f	g	ğ	h	i	ı	j	k	l	m	n	o	ö	p	r	s	ş	t	u	ü	v	y	z

(letters in addition to English: Ç/ç Ğ/ğ İ/ı Ö/ö Ş/ş Ü/ü)

b)Diacritics that can be added to individual letters (these will be dead keys): 
^ - used on âîû (lower and upper cases)

c)Other English letters on the keyboard
Q W X

Total: 33 keys

**4.Word list:** https://raw.github.com/maidis/mythes-tr/master/veriler/kelime-listesi.txt - 88.7k words

**5.Branch:** Turkish branch https://github.com/OptiKey/OptiKey/tree/Turkish

**Comment by feldrim** [comment]
How can we help?

**Comment by JuliusSweetland** [comment]
@feldrim I am working to finish Greek at the moment, so the most useful thing would be to sort out a word list for Turkish. This one (64,000 entries) has been proposed: https://github.com/mertemin/turkish-word-list/blob/master/words.txt

Could you review it and let me know if it will be sufficient and is of good quality?

Also are there any special rules/gotchas with Turkish that are unique to the language? Otherwise I will assume that I just need to create a similar keyboard to the image in the first post at the top, and then add the ability to combine certain letters with diacritic marks to be able to make the following letters (either in lower or upper case): âçğıİîöşüû

With Greek I am moving away from a stand alone key for each required accented letter, i.e. of Turkish to cover letters âçğıİîöşüû I would previously have added 10 extra keys to the keyboard. Now I am moving towards adding diacritic keys which can be combined with standard letters, so instead you would have a keyboard of undecorated letters plus 6 diacritic dead keys. However, it looks like a normal Turkish keyboard has dedicated keys and more keys per row (e.g. 12 keys in the first 2 rows where english has only 10 and 9). Can you comment on what a Turkish user would expect so I don't change the layout too dramatically, or slow down text entry, as a diacritic dead key requires 2 key presses to attach it to a letter, whereas a dedicated accented letters (for example) is only 1.

**Comment by feldrim** [comment]
Well, it is complicated because spelling of Turkish words can be changed via postfixes due to ease of pronunciation. Like when you say **to the street** in Turkish, the word **sokak** (_street_) turns into **sokağa**. The letter **k** at the end turns into **ğ** to make it easy to pronounce. This is a reason when you add a word to the database, you need to add other variations as well. So that almost half of the turkish words get changed with a postfix, by changing the last consonant or adding a consonant after the vowel at the end to make reading two consewutive vowels easily. There will always be some missing words however it is easy to contribute on Github for improvements on the list.

I'd like to inform you about [another wordlist on Github ](https://raw.github.com/maidis/mythes-tr/master/veriler/kelime-listesi.txt). It was a part of [a thesaurus project ](https://github.com/maidis/mythes-tr)for open source grammar checking and other applications. It contains 88.705 words. Since both may contain same words with a few -maybe a few thousands- differences, it would be easy to get a diff-check on two text files and create a better one. Both may be valid but I prefer the latter, since it is not an automatically Wikidictionary-parsed list but an academically curated one.

About the layout, general use is this one, [ASCII 179](http://ascii-table.com/keyboard.php/179). You can find **çğıİöşü** on it but not the ones with **^**. That is because those with **^** like **â** and **î** not considered as a standalone letter but same letter with _a-little-different-pronounciation_. On physical keyboards, we use **shift+3** first, then type the **letter a or i**. Android and iOS keyboards use a different way. When you long press a or i you can see â and î. It is a global thing for touchtype keyboards. You may use whichever vay you want. But if it would be a problem, remember it is just a pronunciation helper and people will get the words without using the ^ sign.

**Comment by JuliusSweetland** [comment]
@feldrim Great info, thank you. I am updating the top post (above). Can you please review.

The only issue I can see is that the dotted capital İ and lower case i are the same letter, as are dotless capital I and lowercase ı (from https://en.wikipedia.org/wiki/Diacritic#Languages_with_letters_containing_diacritics), but the keyboard layout ASCII 179 has the dotted capital İ and lowercase ı on the same key and lower case i and dotless capital I on the same key. Is this really how it is as it would be much easier if the same letters were on the same keys.

**Comment by feldrim** [comment]
The dotless one (upper **I**, lower **ı**) is between **u** and **o** as in standard qwerty. The dotted one (upper **İ**, lower **i**) is the one on the rightmost, just under the **ğ**. On this layout, it is just a matter of using Caps Lock or Shift key to type uppercase or lowercase. There's no difference on upper/lower changing between other letters and **I/ı & İ/i**.

I checked the info on the top. I haven't seen any poroblem with it. That looks awesome.

BTW, I'd love to recommend you an article, [What I learned about languages just by looking at a Turkish typewriter](https://medium.com/@mwichary/what-i-learned-about-languages-just-by-looking-at-a-turkish-typewriter-fc840aab1b0a#.247bjwr4y)  by [Marcin Wichary](https://medium.com/@mwichary). It's not directly related but would be a good read.

**Comment by JuliusSweetland** [comment]
@feldrim Ahh, my mistake. I was using this link http://ascii-table.com/keyboard.php/179 - when you hover over each key it shows you the lower and upper characters. I must have mis-read the two i keys as I thought they were swapped. All good. I'll get to Turkish in 3 versions time (next release is Greek, then a performance fix and a few other things, then Turkish, then Spanish). Please let me know if you require Turkish with more urgency than that.

Also, thanks for the link - I'll read it tomorrow.

**Comment by JuliusSweetland** [comment]
@feldrim In the mean time there are some strings still outstanding on the Transifex site: https://www.transifex.com/optikey/optikey/

**Comment by feldrim** [comment]
You are very welcome. I'm pretty sure planned time to release will help us to improve translations and solve any other problems. No need to rush. I'll have look at it now but since I have an exam on Java and OOP tomorrow, I won't be able to contribute on untranslated strings. But I'll do whatever I can asap.

**Comment by JuliusSweetland** [comment]
Good luck for your exam. Focus on that!

**Comment by feldrim** [comment]
![Image](https://cloud.githubusercontent.com/assets/15658960/13412275/b9a1eae4-df4a-11e5-9a1f-bbba3fd909cc.jpg) I can't join the team because it says "Resource Breakdown". And also when I click on **join team**, I can't find Turkish. The language list end on Thai although I can view the existing members.

**Comment by JuliusSweetland** [comment]
Hmm, is your username Zafer? I just accepted a request from that user for
Turkish. If that's you refresh the page and see. The Transifex site can be
a bit odd also, so a refresh might be worthwhile anyway.

If that doesn't help post you're Transifex username and I'll invite you
manually.

**Comment by feldrim** [comment]
As soon as I commented on this issue, the acceptance mail had come. So I deleted sinc there was no need for that. Thanks.

**Comment by JuliusSweetland** [comment]
@feldrim I read the article - very interesting. Before any work is done, is the QWERTY layout commonly used in Turkey, vs the F layout? I could do either, or both?

**Comment by feldrim** [comment]
Only the former generations who had used typewriters daily are accustomed to the F keyboard. Windows, Linux and Mac has support for it. However mobile OSs and applications nowadays do not support F keyboard, which is not needed anymore. Adding the [F Keyboard](http://ascii-table.com/keyboard.php/440) would be nice but not an obligation. It may be considered in latter releases.

**Comment by JuliusSweetland** [comment]
Understood thanks. Good luck for your exam.

**Comment by feldrim** [comment]
Oh, thanks. I've just came home from the exam. Thank you for your kindness.

**Comment by feldrim** [comment]
What do you plan for Turkish TTS?

**Comment by JuliusSweetland** [comment]
If you are asking about a Turkish TTS engine, does the Windows Turkish
language pack not include a Turkish voice?

I don't supply language specific voices with OptiKey - I make the
assumption that users can source their own voices.

**Comment by feldrim** [comment]
I haven't found it on both Windows 7 and Windows 10. I think additional software is needed. I'll take a look at it.

**Comment by JuliusSweetland** [comment]
Someone recommended this site to me:
https://www.freedomscientific.com/Downloads/synthesizers

I think you select the RealSpeak Solo Direct synthesiser and then Turkish.
I cannot test as it doesn't work on Windows 8.1, but apparently it does on
Windows 7. If you don't have any luck try the Vocaliser Direct downloads
(which include both male and female Turkish voices)

**Comment by feldrim** [comment]
Hi. There's only 9 word/terms last on transifex. Though I don't have proper equipment. So I don't know what those commands do. Translating them word by word would cause wrong translation. I'll take a deeper look in documentation for the issue. Beyond that I'd love to help any other translations like web site, documentation or video subtitling. Nowadays I plan to translate the documentation after finishing that last part of UI translation in transifex.

**Comment by JuliusSweetland** [comment]
I only have access to email right now, but if you post the remaining
English strings I'll explain them for you.

**Comment by feldrim** [comment]
Sorry, I've been late. Here are the last 9 terms I've been stuck:

- CONVERSATION ALPHA
- CONVERSATION NUMERICS & SYMBOLS
- DOCK ACTION KEYS:
- Overrides for individual key fixation times (ms):
- MODIFIER KEYS:
- MOUSE (DO AT CURRENT LOCATION) KEYS:
- MOUSE (DO AT POINT) KEYS:
- MOUSE (MISCELLANEOUS) KEYS:
- Specify times to complete on individual keys:

**Comment by JuliusSweetland** [comment]
Translations below each one...

* CONVERSATION ALPHA
This is the name of the Alpha (the main keyboard containing all the letters) keyboard when you are in CONVERSATION mode, i.e. it is the full screen letters keyboard when you only want to generate speech.

* CONVERSATION NUMERICS & SYMBOLS
As above, but this is the conversation keyboard (fullscreen) which contains numbers and symbols.

* DOCK ACTION KEYS:
The set of keys which perform actions on the docked state of OptiKey. Currently this is only "Expand dock" and "Collapse dock" to change the size of OptiKey when docked.

* Overrides for individual key fixation times (ms):
When the user chooses to specify how long it takes to select a key (using a fixation / dwell trigger), specified on each individual key. There is a default value specified and then you can override (change) this value on specific individual keys, e.g. the default may be 1 second, but you can override the QUIT key to require a 2 second fixation before it is selected to prevent accidental selections.

* MODIFIER KEYS:
Keys which modify other keyboard presses, i.e. SHIFT, CTRL, WIN and ALT. They are called modifiers as pressing P would type a P, but pressing CTRL then P will generally modify the resulting action and mean print in most applications.

* MOUSE (DO AT CURRENT LOCATION) KEYS:
The set of keys which perform a mouse action (e.g. left click) where the mouse cursor is now, i.e. these actions do NOT require the user to select a point to move the cursor to before the action is performed. 

* MOUSE (DO AT POINT) KEYS:
The alternative to the above - these require the user to select a point, the cursor is then moved to that point, and THEN the action (e.g. left click) is performed. 

* MOUSE (MISCELLANEOUS) KEYS:
This is the title for the other mouse actions, e.g. MAGNETIC CURSOR, MAGNIFIER (on/off) etc. It's where mouse actions which don't fit in "DO AT POINT" or "DO AT CURRENT LOCATION" go.

* Specify times to complete on individual keys:
This is a YES/NO setting that specifies whether you want to override the selection time (fixation/dwell) on individual keys. If you say NO then all keys take the same default amount of time to selection. If you specify YES then you see the section with the title "Overrides for individual key fixation times (ms):" (as above) and you can then select selection times on individual keys.


N.B. The strings above which are in capitals are titles used for groups of keys in the management console so the translations should also be in capitals and end in a colon. That's also true for the other new strings that you haven't listed here (which are in capitals).

**Comment by feldrim** [comment]
All words in Transifex are translated. They may need a review, maybe by those who have already made contributions before me.

**Comment by JuliusSweetland** [comment]
Good stuff, thank you

**Comment by feldrim** [comment]
You're very welcome.

**Comment by JuliusSweetland** [comment]
@feldrim FYI Turkish is underway. It won't be released until after Greek, but we can test soon. I'll update you as soon as I have something for you.

**Comment by feldrim** [comment]
Thanks. By the way, there's an incorrect translation in transifex after today's update.

`Turkish (Turkey) / Türk (Türkiye)` must be `Turkish (Turkey) / Türkçe (Türkiye)`

**Comment by JuliusSweetland** [comment]
@feldrim Good catch - corrected now.

**Comment by JuliusSweetland** [comment]
Hmm, we have a complication. The Turkish i's are not going to work with OptiKey as it currently stands.

This is how switching the casing of the i's works in windows and C#:
i (lowercase i) -> I (uppercase i) *incorrect for Turkish*
I (uppercase i) -> i (lowercase i) *incorrect for Turkish*
ı (lowercase dotless i) -> I (uppercase i) *correct for Turkish*
İ (uppercase dotted i) -> i̇ (lowercase i) *correct for Turkish*

Unfortunately OptiKey relies on converting the case of letters using windows/C# logic.

Off the top of my head the following issues exist:
* Any dictionary entries which include a lowercase i will be converted to an uppercase dotless I if that letter is converted to uppercase based on the shift key state.
* Multi-key selections and auto-complete matching will not correctly match a captured (or typed) lowercase i to an uppercase dotted İ and will instead attempt a match using an uppercase dotless I.
* Pressing or locking down shift and selecting the lowercase dotted i will result in an uppercase I, rather an a uppercase dotted İ.

Any words in the TurkishTurkey.dic dictionary file which include an uppercase dotted İ should be fine as this will be preserved (as OptiKey works on the basis that all dictionary entries are lowercase unless uppercase letters mean something, e.g. a name, so it leaves them alone).

**Comment by JuliusSweetland** [comment]
Latest version of Turkish uploaded to "Turkish" branch if you want to check out the problem. I think everything else is complete. All ideas and help is very welcome.

**Comment by JuliusSweetland** [comment]
Actually I can probably just set the desired culture on all usages of ToUpper and ToLower  from the relevant language setting.

This link specifically references the English/Turkish I problem in the example:
https://msdn.microsoft.com/en-us/library/24kc78ka(v=vs.110).aspx

**Comment by JuliusSweetland** [comment]
@feldrim As Q W X are not part of the Turkish alphabet, can I remove them completely? I presume they are only included on the keyboard so people can type words in other languages, but within OptiKey you can change the whole keyboard layout by changing your input language, rendering them unnecessary. What do you think?

**Comment by feldrim** [comment]
1. In Java, we use the Locale object. The usage is like:
`Assert.assertEquals("title", "TITLE".toLowerCase(Locale.ROOT));`
Wth the `Locale.ROOT`, we can set the default lowercare/uppercase conversion according to the system default. I'm pretty sure Microsoft has their own format -like the `culture`. So that you won't have any concerns about `toLower/toUpper`.

2. Removing them would change the whole layout, which would affect the user interaction in a bad way. Since early 80's, Turkish Q keyboard layout has been the standard on all platforms. Almost everywhere on web, people will need those letters. So, it would be a bad design decision by both changing user interface to an unknown layout and making the necessary keys out of sight.

**Comment by JuliusSweetland** [comment]
@feldrim Regarding the layout - if you *never* use letter Q W X in Turkish then they really shouldn't be there as they will fail to match with any words in the Turkish dictionary. Are they only there to be able to type in other latin based languages, i.e. never for Turkish? If so I would like to consider a layout change as it may not be as dramatic as you think, i.e. it's 2 letters removed from the left edge of the top line and 1 near the left edge of the bottom. I am confident I could keep the balance of the keyboard with those changes. Thoughts?

**Comment by feldrim** [comment]
As you've noticed, Turkish Q layout is ineffective for Turkish language. Not only for presence of nonexistent letters like Q, W, X, but also the positioning frequently used letters like ü,ğ,i,ş,ç,ö on the rightmost, which makes users type with the little finger. Though the F Turkish layout is the most effective one known for the language, it's not common. People are accustomed to the Q layout, mostly because all the web requires the letters of English Alphabet. If you believe Optikey is based on the communication via text-to-speech and the web usage is not the main theme, then it's ok.

**Comment by JuliusSweetland** [comment]
I'll leave it as the Q layout for the moment and we can consider the layout
again during testing.

**Comment by JuliusSweetland** [comment]
Turkish pre-release is ready for testing!

While we wait for v2.3.1 (Greek and features) I have created another pre-release with Turkish support: v2,3.2 here: https://github.com/OptiKey/OptiKey/releases

PLEASE NOTE THAT THIS IS NOT A FINAL RELEASE AND SHOULD NOT BE INSTALLED UNLESS YOU ARE BETA TESTING THIS VERSION!

Once installed you will have to change the keyboard language (and optionally the UI language) to Turkish either from the Menu within OptiKey, or from the "WORDS" tab in the management console. All feedback is welcome and appreciated.

**Comment by JuliusSweetland** [comment]
@feldrim Have you had a chance to test the Turkish pre-release (details in my previous post above)?

**Comment by feldrim** [comment]
I'm sorry, it's been a bust week. I've tried. The keyboard seems great. Though I couldn't change the interface language. The program works with English interface + Turkish Keyboard.

**Comment by JuliusSweetland** [comment]
Ok I'll take a look.

**Comment by JuliusSweetland** [comment]
@feldrim I couldn't replicate your problem. I was able to change both the keyboard and interface language to Turkish. Please be aware that the "UI language" is a separate setting.

**Comment by JuliusSweetland** [comment]
@feldrim New pre-release of Turkish (v2.3.2) available here: https://github.com/OptiKey/OptiKey/releases

Can you please give me your feedback as I think everything is ready to be released if you (and anyone else who would like to test) are happy.

**Comment by JuliusSweetland** [comment]
Released in v2.3.2
